test/e2e/service.go

```diff
@@ -1152,6 +1152,53 @@ var _ = framework.KubeDescribe("Services", func() {
 			jig.TestHTTPHealthCheckNodePort(publicIP, healthCheckNodePort, "/healthz", expectedSuccess)
 		}
 	})
+
+	It("should only allow access from service loadbalancer source ranges [Slow]", func() {
+		// this feature currently supported only on GCE/GKE/AWS
+		framework.SkipUnlessProviderIs("gce", "gke", "aws")
+
+		loadBalancerCreateTimeout := loadBalancerCreateTimeoutDefault
+		largeClusterMinNodesNumber := 100
+		if nodes := framework.GetReadySchedulableNodesOrDie(c); len(nodes.Items) > largeClusterMinNodesNumber {
+			loadBalancerCreateTimeout = loadBalancerCreateTimeoutLarge
+		}
+
+		namespace := f.Namespace.Name
+		serviceName := "lb-sourcerange"
+		jig := NewServiceTestJig(c, serviceName)
+
+		By("Prepare allow source ips")
+		// prepare the exec pods
+		// hostPod and acceptPod are allowed to access the loadbalancer
+		hostPod := framework.LaunchHostExecPod(c, namespace, "hostexec-accept")
+		acceptPodName := createExecPodOrFail(f.Client, namespace, "execpod-accept")
+		dropPodName := createExecPodOrFail(f.Client, namespace, "execpod-drop")
+
+		hostPod, err := c.Pods(namespace).Get(hostPod.Name)
+		Expect(err).NotTo(HaveOccurred())
+		accpetPod, err := c.Pods(namespace).Get(acceptPodName)
+		Expect(err).NotTo(HaveOccurred())
+
+		By("creating a pod to be part of the service " + serviceName)
+		// This container is an nginx container listening on port 80
+		// See kubernetes/contrib/ingress/echoheaders/nginx.conf for content of response
+		jig.RunOrFail(namespace, nil)
+		// Create loadbalancer service with source range from node[0] and podAccept
+		svc := jig.CreateTCPServiceOrFail(namespace, func(svc *api.Service) {
+			svc.Spec.Type = api.ServiceTypeLoadBalancer
+			svc.Spec.LoadBalancerSourceRanges = []string{hostPod.Status.HostIP + "/32", accpetPod.Status.PodIP + "/32"}
+		})
+		svc = jig.WaitForLoadBalancerOrFail(namespace, serviceName, loadBalancerCreateTimeout)
+		jig.SanityCheckService(svc, api.ServiceTypeLoadBalancer)
+
+		By("check reachability from different sources")
+		svcIP := getIngressPoint(&svc.Status.LoadBalancer.Ingress[0])
+		checkReachability(true, namespace, hostPod.ObjectMeta.Name, svcIP)
+		checkReachability(true, namespace, acceptPodName, svcIP)
+		// expect not to be reachable
+		checkReachability(false, namespace, dropPodName, svcIP)
+
+	})
```

Comment:
this can leak loadbalancers if the ns is responsible for deleting the svc and jenkins shutsdown the cluster soon after. you need to wait for the lb to disappear eg: https://github.com/kubernetes/kubernetes/pull/33957/files#diff-20a4e2095b63ecd60dd25e78bcd67372R1100